Avoid sending non-numeric floats in cloud foundry integrations

jsoriano opened the pull request:
## What does this PR do?

Cloud Foundry integrations are sending some values as they are received
from the Firehose, some of these values can be floats with non-numeric
values (NaN/Inf), that are not supported by JSON and Elasticsearch.
Add defensive code to avoid sending these values to the outputs.

## Why is it important?

Sending non-numeric float values to some outputs produces errors.

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] ~~I have made corresponding changes to the documentation~~
- [x] ~~I have made corresponding change to the default configuration files~~
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Related issues

- Fixes #22599

elasticmachine commented:
## :green_heart: Build Succeeded


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-22634/detail/PR-22634/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-22634/detail/PR-22634/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-22634/detail/PR-22634/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-22634/detail/PR-22634/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://beats_22634.docs-preview.app.elstc.co/diff)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats

* Build Cause: `[Pull request #22634 opened]`

* Start Time: 2020-11-17T18:00:41.905+0000

* Duration: 47 min 25 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 588  |
| Skipped      | 48 |
| Total        | 636   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY-->


### Steps errors [![1](https://img.shields.io/badge/1%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-22634/detail/PR-22634/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>

  
  
  
  
  
#### `Terraform Apply on x-pack/metricbeat/module/aws`
<ul>
<li>Took 3 min 16 sec . View more details on <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-22634/runs/1/steps/910/log/?start=0">here</a></li>

</ul>
  
  </p>
  </details>




## :green_heart: Flaky test report
Tests succeeded.


<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

### Test stats :test_tube:
| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 588  |
| Skipped      | 48 |
| Total        | 636   |



</p>
</details>



<!--COMMENT_GENERATED_WITH_ID_comment.id-->

elasticmachine commented:
Pinging @elastic/integrations-platforms (Team:Platforms)

ycombinator commented on x-pack/metricbeat/module/cloudfoundry/value/value.go:
```diff
@@ -41,5 +42,24 @@ func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+			r.logger.Debugf("Ignored event with value that is not a number on field %s: %v", name, v)
+			return true
+		}
+	}
```
Should we check if `err != nil && err != ErrKeyNotFound`?

jsoriano replied:
Same as with https://github.com/elastic/beats/pull/22634/files#r528651812, what would we do if this happens?

ycombinator commented on x-pack/metricbeat/module/cloudfoundry/container/container.go:
```diff
@@ -41,5 +42,19 @@ func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+		if v, ok := v.(float64); ok && (math.IsNaN(v) || math.IsInf(v, 0)) {
+			event.RootFields.Delete(cpuPctKey)
+		}
+	}
```
Should we check if `err != nil && err != ErrKeyNotFound`?

jsoriano replied:
What would we do in this case? I only want to remove the field if the key is found (`err == nil`), and the value is a float with non-numeric value.
If there is any error getting the value, or it is not a float, it is ignored. (This shouldn't ever happen, but if it happens, still the value wouldn't be producing problems).

ycombinator replied:
I guess my concern is that if we're expecting there to be a value for this key but it's not there, that's unexpected behavior so something's wrong and maybe we should know about it? If you feel that's too strict, I'm okay leaving this as-is.

jsoriano replied:
Ok, I have refactored the code to log at the debug level the errors found getting the values.
I have also moved the code for the check so it is shared between both metricsets.
Let me know what you think.

ycombinator commented on x-pack/metricbeat/module/cloudfoundry/value/value.go:
```diff
@@ -41,5 +42,24 @@ func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	nameKey := "cloudfoundry.value.name"
+	if v, err := event.RootFields.GetValue(valueKey); err == nil {
+		if v, ok := v.(float64); ok && (math.IsNaN(v) || math.IsInf(v, 0)) {
+			name, _ := event.RootFields.GetValue(nameKey)
```
Should we do any error checking here?

jsoriano replied:
Ok, I will produce a different message if the name is not found.

ycombinator reviewed (commented):
The interfaces to enable mocking for unit testing are 👍👍.  

Just left a couple of questions around error checking.

ycombinator reviewed (approved):
LGTM

jsoriano commented:
@ycombinator leveraging that I moved the common code to a helper function, I have added a unit test for it, please take a quick look after this change. Thanks!

ycombinator reviewed (approved):
Still LGTM.
